bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -300,6 +302,82 @@ void PreProcessor::onMessage<ClientPreProcessRequestMsg>(ClientPreProcessRequest
   handleSingleClientRequestMessage(move(clientMsg), false, 0);
 }
 
+// Should be called under reqEntry->mutex lock
+bool PreProcessor::isRequestPreProcessingRightNow(const RequestStateSharedPtr &reqEntry,
+                                                  ReqId reqSeqNum,
+                                                  NodeIdType clientId,
+                                                  NodeIdType senderId) {
+  if (reqEntry->reqProcessingStatePtr) {
+    const auto &ongoingReqSeqNum = reqEntry->reqProcessingStatePtr->getReqSeqNum();
+    const auto &ongoingCid = reqEntry->reqProcessingStatePtr->getReqCid();
+    LOG_DEBUG(logger(),
+              KVLOG(reqSeqNum, clientId, senderId)
+                  << " is ignored:" << KVLOG(ongoingReqSeqNum, ongoingCid) << " is in progress");
+    preProcessorMetrics_.preProcReqIgnored.Get().Inc();
+    return true;
+  }
+  return false;
+}
+
+// Should be called under reqEntry->mutex lock
+bool PreProcessor::isRequestPassingConsensusOrPostExec(SeqNum reqSeqNum,
+                                                       NodeIdType clientId,
+                                                       NodeIdType senderId,
+                                                       const string &cid) {
+  // Count requests arrived through non-primary replicas and accepted in spite of the
+  // same request has been processing by the primary replica
+  if (clientId != senderId && myReplica_.isClientRequestInProcess(clientId, reqSeqNum)) {
+    preProcessorMetrics_.preProcReqForwardedByNonPrimaryNotIgnored.Get().Inc();
+    LOG_DEBUG(logger(),
+              "Not ignoring client request arrived through a non-primary replica"
+                  << KVLOG(cid, reqSeqNum, clientId, senderId));
+  }
+  // Verify that a request is not passing consensus/PostExec right now. The primary replica should not ignore client
+  // requests forwarded by non-primary replicas, otherwise this will cause timeouts caused by missing PreProcess
+  // request messages from the primary. TBD YS: save senderId in the header and set it before re-sending
+  if ((clientId == senderId) && myReplica_.isClientRequestInProcess(clientId, reqSeqNum)) {
+    LOG_DEBUG(
+        logger(),
+        "The specified request has been processing right now - ignore" << KVLOG(cid, reqSeqNum, clientId, senderId));
```

Comment:
"... is in consensus or being post-executed right now"?

Reply:
I'll change the related logic in the next PR